gix-diff/src/blob/pipeline.rs

```diff
@@ -533,6 +541,10 @@ impl Driver {
     pub fn prepare_binary_to_text_cmd(&self, path: &Path) -> Option<std::process::Command> {
         let command: &BStr = self.binary_to_text_command.as_ref()?.as_ref();
         let cmd = gix_command::prepare(gix_path::from_bstr(command).into_owned())
+            // This program intentionally doesn't get the whole Git context.
+            // In fact, when checked it only saw the `GIT_EXEC_PATH` and `COLORTERM` environment variables, nothing more.
+            // We by default let it inherit the entire environment. On Windows, Git seems to add the env though.
```

Comment:
@EliahKagan I just pushed this intermediate commit to share this comment with you. I just noticed that `gix-command` may provide too much of the parent process environment to the spawned commands, and thus is less hygienic than Git seems to be.
Maybe attackers can abuse this, or even if it's not an attacker, malevolent players of developer software might use it to syphon off information from developer machines that Git doesn't provide to them.

My feeling is the `gix-command` should run `env_clear()` beforehand and then selectively re-add variables to match Git. That, however, probably has to be checked on a case-by-case basis.

Reply:
The merge driver program, similar to the diff-driver (which isn't the `textconv` program) seem to see the inherited environment + the Git context. So it looks like the textconv is special, but technically still sees too much of the environment right now when invoked from `gix-diff`.

Reply:
What sanitization does `git` do here that we are missing? At least with `diff.*.command` and `diff.*.textconv`, I have not been able to observe any removed environment variables. [I did not find the environment to be sanitized on any platform.](https://gist.github.com/EliahKagan/31517ea8d4bb0300a521eb63c6b2efd2) As shown in that gist, I ran `printenv`:

- Directly.
- Through a custom Git command.
- Through a custom Git diff script.
- Through a custom Git textconv script.

All environment variables, whether Git-related or not, seem to be kept. This was never limited to `GIT_EXEC_PATH` and `COLORTERM`. Several have different values and some new variables are set, especially with Apple Git on macOS, but I did not observe any to be removed.

This was the case even when I set `GIT_DIR` and `GIT_WORK_TREE`, which are repository-specific and which are among the environment variables `git` clears when performing an operation on a different repository. However, I did not do that test on all platforms; in particular, I didn't get a chance to do it on macOS.

For simplicity, I did not test with explicitly specified merge tools, since from your description it sounds like diff and merge tools are expected to be treated similarly with respect to environment handling, while textconv tools might be treated differently.

There are [a number of other variables](https://github.com/git/git/blob/57974d46a4d5a079471c4f7eaa5933b1922299e4/environment.c#L128-L148), some [repository-specific](https://github.com/git/git/blob/57974d46a4d5a079471c4f7eaa5933b1922299e4/environment.c#L128-L148) and others not, that Git treats specially and I wouldn't be surprised if it unset them under some circumstances. But before testing with them, I figured I'd ask for clarification.

> or even if it's not an attacker, malevolent players of developer software might use it to syphon off information from developer machines that Git doesn't provide to them

I'm not sure what kind of threat you're describing here. Certainly if one expects some kind of sanitization to occur and it does not, then one might end up causing sensitive information to become available to a component one wishes to limit, or causing untrusted information *from* a component one wished to be limited to be used where it might be wrongly trusted. But that is completely general. If you are talking about malevolent development tools that have access to repositories or can run arbitrary commands, then I don't think this would affect their abilities.

I do recommend that any differences in sanitization from what `git` does, especially where `git` removes an environment variable and we do not, be considered potential vulnerabilities and investigated. But I haven't been able to observe that with diff and textconv tools.

> My feeling is the `gix-command` should run `env_clear()` beforehand and then selectively re-add variables to match Git. That, however, probably has to be checked on a case-by-case basis.

Except in circumstances where Git does that, this does not sound like it would necessarily be an improvement. Since the user may configure arbitrary diff and textconv tools, the user may configure a tool that takes configuration from arbitrary environment variables. If a tool uses environment variables that start with a widely recognized prefix, such as `GIT_`, and uses them in a way that is inconsistent with the way they are commonly used, then that tool is broken. Otherwise, though, sanitizing the environment unexpectedly seems like it could harm usability and correctness, and in cases where a less permissive configuration has been overridden using environment variables, may also harm security.

I suspect I may not fully understand what you are proposing, however.

Reply:
Thanks so much for the detailed analysis - I didn't expect this, but am happy about such a large sample size.

It's astonishing that no environment was sanitised, even though the `textconv` tool in my run definitely didn't have any but two environment variables set.
Something very relevant I forgot to share is my using of a very recent Git executable directly built from `master` (c5ee8f2d1c9d336e0a46139bd35236d4a0bb93ff). It might be possible that something changed.
*Actually, scratch that, I just retried with the standard apple git installation, and it's the same result*.

This is the textconv filter program I am using:

```bash
#!/bin/bash

env
pwd
echo $@
```

> I'm not sure what kind of threat you're describing here.

The best description I can come up with is the loss of confidentiality, along the lines of developer tools making an extra buck by selling developer data. But it's true, if the program is running, some additional environment variables probably are the smallest thing one would worry about.

> Except in circumstances where Git does that, this does not sound like it would necessarily be an improvement. 

Thanks for sharing your thoughts. For now I was just passing this concern by you to see what you think, and it's good to see I can also be a little bit too paranoid. However, I'd definitely keep sharing even the smallest concerns, maybe even as practice, to maybe one day write secure code from the start 😅.

Then I will take no action, but instead pass the standard Git context to everything as it seems to have been passed as well in your examples.

Reply:
🤦‍♂️

I think I know now what happened here. This is the output I was looking at.

```
❯ git --no-pager diff HEAD^ HEAD
diff --git a/file b/file
index 1f7391f..67b7f61 100644
--- a/file
+++ b/file
@@ -49,4 +49,4 @@ WEZTERM_CONFIG_FILE=/Users/byron/.wezterm.lua
 GIT_EXEC_PATH=/Applications/Xcode.app/Contents/Developer/usr/libexec/git-core
 COLORTERM=truecolor
 /Users/byron/dev/github.com/Byron/small
-/var/folders/x7/jgy95vjs3v3ffszwn0wg0gz80000gn/T//git-blob-GMpYec/file
+/var/folders/x7/jgy95vjs3v3ffszwn0wg0gz80000gn/T//git-blob-490Xcv/file
```

After all, I wasn't outputting the information to a separate file, but *as* the file to be diffed.

Hence, what I saw was a patch with most information the same, 50 lines of that… my testing is entirely broken, and so are the results.

So it doesn't only inherit the existing env, it also adds the `git-core` tools to the `PATH`.

Reply:
> Thanks so much for the detailed analysis - I didn't expect this, but am happy about such a large sample size.

I usually test with only something like [`git-run`](https://gist.github.com/EliahKagan/31517ea8d4bb0300a521eb63c6b2efd2#file-script-1-git-run-sh), and usually only to observe what environment Git for Windows prepares. This was an opportunity to see, to some extent, how similar the environment for custom Git commands really is to the environment for external tools. So I was also motivated by that goal.

Accordingly, even though the cause of the environment seeming to be sanitized has been discovered, I have gone ahead and repeated the experiment with `GIT_DIR` and `GIT_WORK_TREE` set [on macOS](https://gist.github.com/EliahKagan/31517ea8d4bb0300a521eb63c6b2efd2#macos---with-git_dir-and-git_work_tree-set).

Of course, there are still specific additions and removals that this does not reveal, and some of them may be relevant to compatibility goals for gitoxide. For example, I have separately observed that, at least in recent versions of Git, the `GIT_CONFIG_COUNT` variable is removed from the environment for an external transport command such as `ssh`, even when it is present in the immediate parent of that transport command. I'm not sure why that is done, if it is important, or if `gix-transport` should behave similarly. I imagine there are a number of specific behaviors like this.

For `gix-transport`, it seems [already to remove a number of variables](https://github.com/Byron/gitoxide/blob/ab8880fbdf0f7af7b483f4e1f9adbb8e374183ee/gix-transport/src/client/blocking_io/file.rs#L17-L35). This has long included `GIT_CONFIG_COUNT`, which was added in https://github.com/Byron/gitoxide/commit/aa315b4a136bec57a1cce4b245c606904adc5c12. But I am unsure if it the list there is entirely current.

#### Whether we use a shell or not is significant to compatibility

Maybe a more significant compatibility concern, including for `gix-transport`, is that there seem to be some configurable external commands where whether or not a shell is used to run them is part of the semantics of how they are configured, but gitoxide does not currently honor those semantics.

For example, when `git` runs a SSH client from `GIT_SSH`, it never uses a shell, and when it runs a SSH client command from `GIT_SSH_COMMAND` (or `core.sshCommand`), it always uses a shell. [This is documented](https://git-scm.com/docs/git#Documentation/git.txt-codeGITSSHcode) and I think the `git` documentation intends that users be able to rely not just on `GIT_SSH_COMMAND` using a shell, but also on `GIT_SSH` either not using a shell or behaving impeccably as though it does not. I think that gitoxide not always adhering to this may actually be the cause of #1432.

Even if the effect of shell processing triggered by [special characters](https://github.com/Byron/gitoxide/blob/ab8880fbdf0f7af7b483f4e1f9adbb8e374183ee/gix-command/src/lib.rs#L101) in the original command name is correct, whether a command is run directly or through a shell can have important effects, at least on Windows where whether a command is run directly or through another program can have important effects even when that other program is not a shell.

This is because, on Windows, [programs with different "bit-ness" see some parts of the filesystem differently](https://learn.microsoft.com/en-us/windows/win32/winprog64/file-system-redirector). When `ssh` is `C:\Windows\System32\OpenSSH\ssh.exe` or `GIT_SSH` points there, and `gix` is a 64-bit build but the shell `gix-command` finds is a 32-bit program, then when it uses a shell, it will fail to find a usable `ssh.exe`, because there is no `C:\Windows\SysWOW64\OpenSSH\ssh.exe`.

A shell may be used more often on Windows when the caller permits it, because an environment variable like `GIT_SSH` may contain `\` as a path separator, which [is matched here](https://github.com/Byron/gitoxide/blob/ab8880fbdf0f7af7b483f4e1f9adbb8e374183ee/gix-command/src/lib.rs#L101). This may, in particular, lead to attempts to work around the problem by adding more backslashes, which may sometimes work, since suitably repeated non-leading backslashes cause backslashes to be treated literally by shells while being equivalent to single backslashes when a shell is not used, which is also consistent with the description in #1432.

#### The environment may be relevant to deniability for data exfiltration

> The best description I can come up with is the loss of confidentiality, along the lines of developer tools making an extra buck by selling developer data.

Unless isolated in a way that would presumably also sanitize the environment, I would expect such a tool could obtain that information if it has fairly unrestricted disk access compared to the user running it.

On the other hand, this may be a real concern in that including all environment variables and their values in telemetry or crash reporting data might not be identified as a malicious behavior.

I think the best protection against that, though, is to sanitize the environment before we are ourselves doing anything with it, rather than afterwards, and preferably to sanitize it before we even have access to it. That is, in the best case, when using a program to run tools whose environments one wishes to sanitize, one would sanitize the environment for that program. For example, one might run `gix` or `ein` (or `git`) with a sanitized environment if one wishes to use it to run tools whose environments should be limited.

In particular, there seems to be some overlap in environment variables one might want to sanitize and those one might need for a custom tools. For example, since a textconv tool is [typically lossy](https://git-scm.com/docs/gitattributes#_performing_text_diffs_of_binary_files), it may sometimes make sense for it to use a service that accesses real-time data, or that uses a machine learning model that might be infeasible or inconvenient to run locally. (For example, an image could be presented as an automatically generated caption describing it.) But these are also two kinds of use cases where one often has an API key in an environment variable that one would wish the tool to access and use.

So I don't know how much we can help with this, other than by doing all the sanitization `git` documents it does, or otherwise does, under similar circumstances, and by improving compatibility with sanitization that has already been performed by attempting to avoid malfunctioning even if our own environment is unusually sparse.

#### Miscellaneous

> However, I'd definitely keep sharing even the smallest concerns

Yes, definitely please do!

> After all, I wasn't outputting the information to a separate file, but *as* the file to be diffed.
>
> Hence, what I saw was a patch with most information the same, 50 lines of that… my testing is entirely broken, and so are the results.

Ah, that makes sense. (It was to avoid this sort of confusion that I had [`dumpenv`](https://gist.github.com/EliahKagan/31517ea8d4bb0300a521eb63c6b2efd2#file-script-2-dumpenv-sh) always dump the environment to a file of a fixed name in the current directory.)

> So it doesn't only inherit the existing env, it also adds the `git-core` tools to the `PATH`.

On macOS, at least on the system I tested on, some other environment variables were also added.

(The situation on Windows may be dissimilar and less relevant. On Windows, I tested in in a Git Bash environment so that I could run the exact same commands, and I believe there is some MSYS path normalization logic that affected what `git` itself was seeing, so some of the other changes may not have been done by `git`.)

#### `PATH`, but also `GIT_CONFIG_PARAMETERS`, are significant to compatibility

> It seems that some work will be needed to assure the environments are the same, for best compatibility.

I don't know what we should do, and I think it might even be best to deliberately refrain from doing so by default, maybe with the ability to change this behavior through `gitoxide.*` configuration variables, environment variables, or both.

For prepending the path of the `git-core` directory to `PATH` , I think the disadvantages of us doing that by default are:

1. There is no corresponding directory for a tool that performs Git operations with a `gix*` crate. It will not be "our" `git-core` directory, and the tools in it will not do anything to integrate back with gitoxide.
2. There are some guarantees relevant to safety that Git can provide and we cannot. Because we may be running on a system that does not have Git installed, or where we have not found Git, we cannot always prepend a `git-core` directory to `PATH`. But developers will likely have Git installed with `git` in their `PATH`, so this will make it very easy to write bugs that assume `PATH` starts with a `git-core` directory.
3. The consequences of guessing the wrong `git-core` directory may be worse than those of reading "installation" configuration from a different `git` installation than the one currently being used. In particular, guessing the wrong "installation" configuration is likely to be discovered sooner than guessing the wrong `git-core` directory, whose effects are minimal until a subprocess run via `gix-command` runs `git` or commands found only in `git-core`. Then again, this may be possible to avoid by unifying the way we determine both locations.
4. We do not currently support other environment changes made by `git`. We may be called *by* `git`. This is relevant both when we when have a descendant `git` process and when don't.

Regarding point 4, I am aware of two cases I believe to be significant:

- One is a a special case of the above point 3 above. If we prepend a `git-core` directory to `PATH` based on our search for `git`, we need to make sure we always find the `git` that called us. Ordinarily this would happen automatically, but I think it could go wrong in some cases with the `EXEPATH` optimization on Windows.

- Another case is when we are called in a custom `git` command, or otherwise as a tool used by `git`, and the `-c` option has been passed to `git` to set command scope configuration. `git` uses the `GIT_CONFIG_PARAMETERS` environment variable to pass such configuration to subprocesses--both those that are conceptually part of `git`, and custom scripts. With [that `git-run`](https://gist.github.com/EliahKagan/31517ea8d4bb0300a521eb63c6b2efd2#file-script-1-git-run-sh):

  ```text
  ek@Glub:~$ git -c foo.bar=baz run printenv | sort | grep ^GIT
  GIT_CONFIG_PARAMETERS='foo.bar'='baz'
  GIT_EXEC_PATH=/usr/lib/git-core
  ```

  But `gix-config` does not recognize `GIT_CONFIG_PARAMETERS`, as can be verified by running `git -c foo.bar=baz run gix config` in a repository and inspecting the output.

Then again, point 4 may be less of a disadvantage than an opportunity:

- If `PATH` starts with a `git-core` directory, and inspection of the directory reveals it has the expected contents, then that could probably be used to infer the same information we get on Windows from `EXEPATH` when `EXEPATH` is defined. Checking this before checking `EXEPATH` would reduce the risk of guessing wrong based on `EXEPATH`.

- Maybe gitoxide should recognize `GIT_CONFIG_PARAMETERS`. It seems to me that this would be a nice feature. But it should possibly be opt-in, since while the parsing rules for `GIT_CONFIG_PARAMETERS` will *probably* not change in an incompatible way, it is really an implementation detail.

  ([`GIT_CONFIG_{COUNT,KEY,VALUE}`](https://git-scm.com/docs/git-config#Documentation/git-config.txt-GITCONFIGCOUNT) was introduced in part because the parsing rules for `GIT_CONFIG_PARAMETERS` are nonobvious and hard to handle manually, rather than out a desire to be able to make incompatible changes to `GIT_CONFIG_PARAMETERS`. But [the commit](https://github.com/git/git/commit/d8d77153eafdb0fc334e827976f09e4bdff26b58) that introduced `GIT_CONFIG_{COUNT,KEY,VALUE}` noted that `GIT_CONFIG_PARAMETERS` is "for internal use only," which seems still to be the case since it remains [undocumented](https://git-scm.com/docs/git-config#ENVIRONMENT) in [git-config(1)](https://git-scm.com/docs/git-config) and [seems not](https://git-scm.com/search/results?search=GIT_CONFIG_PARAMETERS) to be mentioned at all in any of the Git documentation.)

  That it is an implementation detail suggests we should not use it in any way, even to obtain configuration from it when present. But if we don't obtain configuration from it, then we ignore command-scope configuration passed via `-c` in a `git` invocation that calls us as a child process, and thus we do not have the *full* interoperability with `git` that prepending a detected `git` installation's `git-core` directory to `PATH` would be done to attempt to achieve.

Maybe gitoxide should have some default level of attempted interoperability with a detected installed `git`--similar to what we do now, looking for and using installation-level configuration but not going much beyond that--and also recognize one or more `GIX_*` environment variables that that allow the level of attempted interoperability to be decreased (not looking for or using anything from the installed `git`), increased (recognizing `GIT_CONFIG_PARAMETERS`, prepending `git-core` to `PATH`), or otherwise adjusted (specifying the path to a `git` installation to be used instead of the current heuristics). Or maybe these should each be individually configured knobs, with no broad "level" aggregating them.

On the other hand, I think there is an argument for reading configuration from `GIT_CONFIG_PARAMETERS` by default, even if none of the rest of these ideas are done. We already look for `git` configuration associated with the installation in ways that are not limited to guarantees provided by Git:

- This is most pronounced in Windows. On Windows, we look for Git in places where it it is installed by default but is not always installed. We assume that if Git is found there and is not in `PATH`, then its configuration should be used. We also avoid some `git` invocations on Windows, where they can be slow, by using information from `EXEPATH`, which I don't think is documented. `EXEPATH` is named generally enough that some other software might be using it incompatibly, though I think the practical risk from that is low.
- But it also applies to other systems. For example, as far as I know, the way `gix-path` obtains installation-level configuration from Apple Git relies on experimentally observed behavior that has not been fully documented. Furthermore, if some other downstream Git distribution added other custom scopes that were lower rather than higher, then it might be broken, even after [#1523](https://github.com/Byron/gitoxide/pull/1523) and [#1567](https://github.com/Byron/gitoxide/pull/1567).

So by gitoxide's current standards of how it obtains configuration, it seems like it should probably obtain configuration from `GIT_CONFIG_PARAMETERS` when present, as well. Furthermore, `GIT_CONFIG_PARAMETERS` is at least somewhat visible. As quoted in #147, changes to it have in the past had Git changelog entries.

Reply:
> For `gix-transport`, it seems [already to remove a number of variables](https://github.com/Byron/gitoxide/blob/ab8880fbdf0f7af7b483f4e1f9adbb8e374183ee/gix-transport/src/client/blocking_io/file.rs#L17-L35). This has long included `GIT_CONFIG_COUNT`, which was added in [aa315b4](https://github.com/Byron/gitoxide/commit/aa315b4a136bec57a1cce4b245c606904adc5c12). But I am unsure if it the list there is entirely current.

Thanks again for researching this, and I glad `gix-transport` isn't entirely off when launching `ssh`. Being 90% there seems like a good start. Assuring that is good enough or even safe will be harder though.

> [..] That is, in the best case, when using a program to run tools whose environments one wishes to sanitize, one would sanitize the environment for that program. For example, one might run `gix` or `ein` (or `git`) with a sanitized environment if one wishes to use it to run tools whose environments should be limited.

That makes sense. However, it's also a stance that `gix` (the crate) itself may take, a configurable choice that is inherited by any program that is using it. In that sense, `git` can reasonably assume that users are expecting the entire environment to be inherited, and maybe `gix` (the CLI) should mirror that behavior. Probably there are also other cases when it's better to pass only the most necessary information.
Part of me thinks that I am too hung up on the confidentiality aspect of it, and that it would be best to just do what Git does and to pass any other handling of the environment to the user of `gix`, maybe even explicitly so by making this concern part of the documentation. After all, `gix` (the crate) can't really offer facilities for this as it's unclear what 'sanitized' has to look like.

> So I don't know how much we can help with this, other than by:
> 
> * Doing all the sanitization Git itself performs or documents as recommended, under similar circumstances.
> * Improving compatibility with sanitization that has already been performed by attempting to avoid malfunctioning even if our own environment is unusually sparse.

This convinces me to only document the behaviour, i.e. the attempt to produce similar environments as Git would, with everything else left to the user of the library.

> I don't know what we should do, and I think it might even be best to deliberately refrain from doing so by default, maybe with the ability to change this behavior through `gitoxide.*` configuration variables, environment variables, or both.

Without having read the remainder of the paragraph yet, my intuitive response is to be as compatible to Git as possible, maybe with a way to opt-out.
The reason for that is that tools that use `gix` expect to perform work on Git repositories, and should work similarly to Git by default. It's a selling point of `gitoxide` as a project - be able to 'be Git' but with your own tools.


> For prepending the path of the `git-core` directory to `PATH` , I think the disadvantages of us doing that by default are:
> 
> 1. There is no corresponding directory for a tool that performs Git operations with a `gix*` crate. It will not be "our" `git-core` directory, and the tools in it will not do anything to integrate back with gitoxide.
> 2. There are some guarantees relevant to safety that Git can provide and we cannot. Because we may be running on a system that does not have Git installed, or where we have not found Git, we cannot always prepend a `git-core` directory to `PATH`. But developers will likely have Git installed with `git` in their `PATH`, so this will make it very easy to write bugs that assume `PATH` starts with a `git-core` directory.
> 3. The consequences of guessing the wrong `git-core` directory may be worse than those of reading "installation" configuration from a different `git` installation than the one currently being used. In particular, guessing the wrong "installation" configuration is likely to be discovered sooner than guessing the wrong `git-core` directory, whose effects are minimal until a subprocess run via `gix-command` runs `git` or commands found only in `git-core`. Then again, this may be possible to avoid by unifying the way we determine both locations.
> 4. We do not currently support other environment changes made by `git`. We may be called _by_ `git`. This is relevant both when we when have a descendant `git` process and when don't.

I see that finding the right `git` executable and `git-core` location would be relevant to correctness, and that it's not trivial to do so in all the cases. Maybe you have ideas on how to collect the knowledge presented here in a tracking issue so we don't loose it? If you say you have it in your mind or in your own notes, that would already be good enough for me, I don't want to cause unnecessary overhead.

> Maybe gitoxide should have some default level of attempted interoperability with a detected installed `git`--similar to what we do now, looking for and using installation-level configuration but not going much beyond that--and also recognize one or more `GIX_*` environment variables that that allow the level of attempted interoperability to be decreased (not looking for or using anything from the installed `git`), increased (recognizing `GIT_CONFIG_PARAMETERS`, prepending `git-core` to `PATH`), or otherwise adjusted (specifying the path to a `git` installation to be used instead of the current heuristics). Or maybe these should each be individually configured knobs, with no broad "level" aggregating them.

I like the idea of planning to increase this kind of compatibility but also to provide configuration for it. The way to do this would be to create a `gitoxide.*` configuration key that has an environment variable associated with it. It is then picked up automatically, written into in-memory git-configuration, and available by the code that would have to prepare the command-context.

> On the other hand, I think there is an argument for reading configuration from `GIT_CONFIG_PARAMETERS` by default, even if none of the rest of these ideas are done. We already look for `git` configuration associated with the installation in ways that are not limited to guarantees provided by Git: [..]

My intuition here, still, is to aim for max compatibility and do as much as possible, even if it means to read `GIT_CONFIG_PARAMETERS` and rely on an implementation detail when trying to decode them.

> So by gitoxide's current standards of how it obtains configuration, it seems like it should probably obtain configuration from `GIT_CONFIG_PARAMETERS` when present, as well. Furthermore, `GIT_CONFIG_PARAMETERS` is at least somewhat visible. As quoted in #147, changes to it have in the past had Git changelog entries.

And that's a great find! The parameter parser [is here](https://github.com/gitext-rs/git-config-env/blob/main/src/param.rs) and it seems it would be straightforward to adapt it to not rely on `String`, and to use `BString` instead.

Reply:
> Maybe you have ideas on how to collect the knowledge presented here in a tracking issue so we don't loose it?

Yes, I will open at least two issues related to this: one issue specifically about `GIT_CONFIG_PARAMETERS`, because there are some further subtleties to iron out for that, and also because I think supporting that would be valuable even independently of these other compatibility considerations; and at least one issue for the other things.

I've also noticed #1432 has been closed. Since my hypothesis about the cause is not confirmed, and there may not be enough information there to know the cause, I am disinclined to reopen that (unless the reporter were to decide to reopen it). Instead, I can either include the shell bug in a combined issue or open one specifically for it, whichever seems best once I've researched it slightly more.

Reply:
A brief note, mostly to myself: When I open the issues mentioned above in https://github.com/GitoxideLabs/gitoxide/pull/1585#discussion_r1764341739, I'll also include something about the environment variable handling compatibility issue mentioned in https://github.com/GitoxideLabs/gitoxide/pull/1620#discussion_r1797634536.